### DISCUSSION TITLE
Use external API with PassportJS

### DISCUSSION THREAD

**Discussion by theprobugmaker** [vercel/next.js#13552] [Help] [answered]
I have an external API with PassportJS authentication configured in it using the local strategy and everything is working perfectly. The question I have is related to `with-passport` example as it's confusing me in a few things.
I noticed that both the client and the actual server is in the same example, even the strategy itself so I was wondering, which portion of the code I can remove and use an external instead? Specially the cookie portion cause everything that the external api returns is a cookie with the session in it and I need to use it to every request made by the application.

**Comment by juanbzz** [comment, marked as answer]
The [`with-passport`](https://github.com/vercel/next.js/tree/canary/examples/with-passport) example is to explain how to use [Passport](http://www.passportjs.org/) with API routes. I think all examples aim to show how to use different tech in Next.js.

So, I think the example won't be of much help when you use an external API with Passport.

There are different ways to approach your setup, you can make calls from the client to your external API and share a token between the Next.js app and your API. You can also create API routes in your Next.js app that will call your external API.

If you need to use an external API and want to keep the session management in Next.js you could handle your sessions with API routes and query your external API from there too.

**Reply by theprobugmaker** [reply]
You are right but I think that handling the session on Next and also on the external API isn't a good option as the session is actually stored in a cookie that is on the browser, perhaps using `SWR` with axios or fetch is possible to send the cookies along with the requests like using `credentials: true`.

My actual problem is protecting the authenticated pages, the `with-passport` example has a profile page but it doesn't "protect" it only redirects the user and it has a problem. The user can actually see the profile page and its content it will only redirect once it's rendered by the browser. It doesn't work like a middleware or intercepts the user before even loading the actual page.

**Reply by juanbzz** [reply]
To protect the restricted pages you run the verification logic on `getServerSideProps` so it will redirect the user before the page renders. 

```js
export const getServerSideProps = async (ctx) => {
  const { user, authorized } = verifyUser(ctx); // You have to write the logic to verify the user.

  return { props: { user, authorized } };
};
```

Here's an example of how I do it:

* https://github.com/j0lv3r4/next-authentication/blob/master/examples/simple/pages/profile-ssr.js

**Reply by theprobugmaker** [reply]
You rock, thank very much.

**Reply by theprobugmaker** [reply]
@j0lv3r4 I just noticed something, isn't it insecure to use `getServerSideProps` for this? For example, in the docs it says that when the user clicks a link and the client-side does a page transition it will call the `getServerSideProps` internally as an API request, what if the user intercepts and modifies `authorized` to `true` and `user` with some information? :thinking: 

@tim-phillips @Timer

**Reply by rafaelalmeidatk** [reply]
Do you mean intercepting the request for the new page and adding your own values? That wouldn't matter since you already know the data you are adding to the response, and even if you sucessfully transition to a protected page you won't have access to any private data because 1. you wrote the data of the page yourself in the response or 2. the request for more protected info (in the client-side) should require authentication

**Reply by theprobugmaker** [reply]
@rafaelalmeidatk Sorry but I didn't understand what you mean with `you already know the data you are adding to the response`. (É Brasileiro? Se sim, preciso muito uma ajuda com isso :cry: )

**Reply by rafaelalmeidatk** [reply]
Lets say you have a button that makes a request to `/api/secretUserRoute`. This request will return if the app should redirect to a private page and some private data this page should display.

This is the structure of the response:
```js
{
  authorized: boolean,
  superSecretInfo: {
    name: string,
    likesPizza: boolean,
  }
}
```

You notice this request and to get access to the page, you intercepet it and set the `authorized` value to `true`. You also need to fill the rest of the data so you use random values:
```js
{
  authorized: true,
  superSecretInfo: {
    name: "abc",
    likesPizza: false,
  }
}
```
Now when you click in the button, the app will receive this response and redirect to the secret page, but what do you get? Nothing that you already didn't know, the page will show the name "abc" but you typed it yourself in the response. Btw I am in Reactiflux if you want to ping me there

**Reply by juanbzz** [reply]
> @j0lv3r4 I just noticed something, isn't it insecure to use `getServerSideProps` for this? For example, in the docs it says that when the user clicks a link and the client-side does a page transition it will call the `getServerSideProps` internally as an API request, what if the user intercepts and modifies `authorized` to `true` and `user` with some information? thinking
> 
> @tim-phillips @Timer

You can verify every request by validating the cookies. If `authorized` is changed to `true` you still have to validate the cookies and that's where you redirect the user when the cookies are invalid. The `authorized` is a convenience for your client's logic, it shouldn't be what defines if the user is really authorized or not.

**Reply by theprobugmaker** [reply]
@rafaelalmeidatk @j0lv3r4 You guys are right, it doesn't make much sense but for some reason I thought it would work manipulate the json sent by the server with the user's info and an authorized flag.

@j0lv3r4 How would you validate a cookie? I mean in my case I just have the cookie with the user's session in it, I would need to be calling the server to check if the user is still logged in there like calling a `me` endpoint for example? Or there's another way?

**Reply by hanselabreu** [reply]
> If you need to use an external API and want to keep the session management in Next.js you could handle your sessions with API routes and query your external API from there too.

Let's say I keep the session management within Next.js and store the http cookies. How can I forward that cookie to an external API on every request on the Next.js server side?

**Reply by juanbzz** [reply]
@hanselabreu You add the cookie to the request headers, e.g., 

```js
const result = await fetch("api/", {
  headers: {
    cookie: 'cookieKey=cookieValue"
  }
});
```

**Reply by hanselabreu** [reply]
@j0lv3r4 well in my case I am using `apollo`. Not sure if you're familiar with it, but my problem is that I can't access the `context` to get the `req.headers` to pass along the `cookies` within the `ApolloClient`. Take a look below:

```
const httpLink = createUploadLink({
  uri: "http://localhost:4000",
  credentials: "same-origin",
})

const authLink = setContext((_, ctx) => {
  const { headers } = ctx // this is wrong, it just returns `undefined`
  const cookies = typeof window === "undefined" ? nextCookie(ctx) : cookie.get()
  return {
    headers: {
      ...headers,
      authorization: concat("Bearer ", cookies?.token || ""),
    },
  }
})

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    link: authLink.concat(httpLink),
    cache: new InMemoryCache(),
  })
}
```

So what I was wondering, is if there's a way to explicitly include all the headers to external API's
